### ISSUE TITLE
A mismatch of loss lunction mentioned in paper？

### ISSUE THREAD

**Issue by Leek-A** [tengyu1998/SCI#10]
```python
def norm(self, tensor, p):
        return torch.mean(torch.pow(torch.abs(tensor), p))
...
pixel_grad1 = w1 * self.norm((self.output[:, :, 1:, :] - self.output[:, :, :-1, :]), p)

```
Since the result of self.norm() is a scalar, the weights can't be done as the same as the equation(5) mentioned in your paper.
Maybe it should be like:
```python
pixel_grad1 = w1 * torch.norm(self.output[:, :, 1:, :] - self.output[:, :, :-1, :], p, dim=1, keepdim=True)
```

**Comment by sky135410** [comment]
> ```python
> def norm(self, tensor, p):
>         return torch.mean(torch.pow(torch.abs(tensor), p))
> ...
> pixel_grad1 = w1 * self.norm((self.output[:, :, 1:, :] - self.output[:, :, :-1, :]), p)
> ```
> 
> Since the result of self.norm() is a scalar, the weights can't be done as the same as the equation(5) mentioned in your paper. Maybe it should be like:
> 
> ```python
> pixel_grad1 = w1 * torch.norm(self.output[:, :, 1:, :] - self.output[:, :, :-1, :], p, dim=1, keepdim=True)
> ```
> 
> Another Problems is：
> 
> ```python
> sigma_color = -1.0 / 2 * self.sigma * self.sigma
> ```
> 
> which should be :
> 
> ```
> sigma_color = -1.0 / (2 * self.sigma * self.sigma)
> ```

I think you are correct. 
But when I using the loss above, the EnhanceNet output will close to it input low-light image, then the final output (input/EnhanceNet(input)) will like over-exposure image. 
I don't know which stage is wrong. Do your training the model with above loss and get resonable results?

**Comment by Leek-A** [comment]
> 
> I think you are correct. But when I using the loss above, the EnhanceNet output will close to it input low-light image, then the final output (input/EnhanceNet(input)) will like over-exposure image. I don't know which stage is wrong. Do your training the model with above loss and get resonable results?

Actually, I haven't trained the network, and I think maybe there is something wrong with the super parameters. Since the author used sigma=0.1, it should be 10 if you changed the sigma_color.

**Comment by Leek-A** [comment]
Here is my implementation of loss function:

- Using kornia.color.rgb_to_ycbcr(img)
- Using torch.norm()
- You can set region size
- You can choose sum mode (mean or sum)

```python
def ColorTVLoss(img, w_target=None, region=5, sigma=1.0):
    """

    :param img: Output of model with [N, C, H, W]
    :param w_target: default=img for calculating weights, can be Output/Input of the model
    :param region: default=5
    :param sigma: default=1.0
    :return:
    """
    if w_target is None:
        w_target = torch.clone(img)

    loss = torch.tensor(0.0)
    img_ycbcr = kornia.color.rgb_to_ycbcr(w_target)
    slicers= _ColorTVLoss_create_slicer(region)  # {'H':(s0, s1), 'W':(s0, s1)}
    for slicer in slicers:
        diff_norm = img[..., slicer['H'][0], slicer['W'][0]] \
                    - img[..., slicer['H'][1], slicer['W'][1]]
        diff_w = img_ycbcr[..., slicer['H'][0], slicer['W'][0]] \
                 - img_ycbcr[..., slicer['H'][1], slicer['W'][1]]
        loss += _ColorTVLoss_single(diff_norm, diff_w, sigma=sigma, sum_mode=torch.mean)
    return loss * 2.0   # it is symmetrical


def _ColorTVLoss_create_slicer(region: int = 5):
    """

    :param region: Must be odd (2k-1)
    :return: slicer list
    """
    m = [(x - (region // 2), y - (region // 2)) for x in range(region) for y in range(region)]
    m = m[len(m) // 2 + 1:]
    dict_list = []
    for item in m:
        a_dict = {}
        key_list = ['H', 'W']
        for i in range(len(item)):
            if item[i] == 0:
                a_dict[key_list[i]] = (slice(None, None), slice(None, None))
            elif item[i] > 0:
                a_dict[key_list[i]] = (slice(item[i], None), slice(None, -item[i]))
            elif item[i] < 0:
                a_dict[key_list[i]] = (slice(None, item[i]), slice(-item[i], None))
        dict_list.append(a_dict)
    return dict_list


def _ColorTVLoss_single(diff_norm, diff_w, sigma=0.1, norm_p=1, sum_mode=torch.mean):
    w_sigma = -1. / (2 * sigma * sigma)
    w = torch.exp(torch.sum(torch.pow(diff_w, 2), dim=1, keepdim=True) * w_sigma)
    norm = torch.norm(diff_norm, norm_p, dim=1, keepdim=True)
    return sum_mode(w * norm)
```

**Comment by tengyu1998** [comment]
Thank you for your interest in this project. You pointed out the problem in the project, we will update it in the code. Thanks again!
